target/linux/ar71xx/files/arch/mips/ath79/mach-ap531b0.c

```diff
@@ -0,0 +1,121 @@
+
+/*
+ * Rockeetech AP531B0 11ng wireless AP board support
+ *
+ *  Copyright (C) 2013 Gabor Juhos <juhosg@openwrt.org>
+ *  Copyright (C) 2016 roboidler <roboidler@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License version 2 as published
+ *  by the Free Software Foundation.
+ */
+
+#include <linux/platform_device.h>
+#include <linux/ath9k_platform.h>
+#include <linux/ar8216_platform.h>
+
+#include <asm/mach-ath79/ar71xx_regs.h>
+
+#include "common.h"
+#include "dev-eth.h"
+#include "dev-gpio-buttons.h"
+#include "dev-leds-gpio.h"
+#include "dev-m25p80.h"
+#include "dev-spi.h"
+#include "dev-usb.h"
+#include "dev-wmac.h"
+#include "machtypes.h"
+#include "pci.h"
+#include "dev-ap9x-pci.h"
+
+
+
+#define AP531B0_GPIO_LED_12		12
+#define AP531B0_GPIO_LED_11		11
+
+
+#define AP531B0_GPIO_RST_BTN		17
+
+#define AP531B0_KEYS_POLL_INTERVAL	20	/* msecs */
+#define AP531B0_KEYS_DEBOUNCE_INTERVAL	(3 * AP531B0_KEYS_POLL_INTERVAL)
+
+#define AP531B0_MAC0_OFFSET		0
+#define AP531B0_MAC1_OFFSET		6
+#define AP531B0_WMAC_CALDATA_OFFSET	0x1000
+
+
+static struct gpio_led ap531b0_leds_gpio[] __initdata = {
+	{
+		.name		= "ap531b0:green:g11",
+		.gpio		= AP531B0_GPIO_LED_11,
+		.active_low	= 1,
+	},
+	{
+		.name		= "ap531b0:green:g12",
+		.gpio		= AP531B0_GPIO_LED_12,
+		.active_low	= 1,
+	}
+};
+
+static struct gpio_keys_button ap531b0_gpio_keys[] __initdata = {
+	{
+		.desc		= "reset",
+		.type		= EV_KEY,
+		.code		= KEY_RESTART,
+		.debounce_interval = AP531B0_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= AP531B0_GPIO_RST_BTN,
+		.active_low	= 1,
+	},
+};
+
+
+
+static void __init ap531b0_gpio_led_setup(void)
+{
+
+	ath79_register_leds_gpio(-1, ARRAY_SIZE(ap531b0_leds_gpio),
+			ap531b0_leds_gpio);
+			
+	ath79_register_gpio_keys_polled(-1, AP531B0_KEYS_POLL_INTERVAL,
+			ARRAY_SIZE(ap531b0_gpio_keys),
+			ap531b0_gpio_keys);
+
+}
+
+static void __init ap531b0_setup(void)
+{
+	u8 *art = (u8 *) KSEG1ADDR(0x1fff0000);
+	u8 *pmac;
+
+	ath79_register_m25p80(NULL);
+	ap531b0_gpio_led_setup();
+	ath79_register_usb();
+
+	ath79_register_mdio(0, 0x0);
+	ath79_register_mdio(1, 0x0);
+
+	pmac = art + AP531B0_WMAC_CALDATA_OFFSET + 2;
```

Comment:
Yes, i`d checked that.
here`s the art mtd:
root@LEDE:/# cat /dev/$(cat /proc/mtd | grep -i art | awk -F':' '{print $1}') | 
hexdump
0000000 ffff ffff ffff ffff ffff ffff ffff ffff
*
0001000 0202 d825 b008 1f3f 0000 0000 0000 0000
0001010 0000 0000 0000 0000 0000 0000 0000 1f00
0001020 3302 0000 0000 0400 2800 4d04 0300 08ff
0001030 2100 0000 0010 0110 0022 2202 0050 0150
0001040 0150 0100 0000 0000 0014 0064 7da4 af00
0001050 ff00 0000 0000 0000 0000 0000 0000 0000
0001060 0e0e 0300 2ce2 0002 0e1c 80c0 800c 80c0
0001070 8000 0000 0000 0000 0000 0000 0000 0000
0001080 0000 0000 0000 0000 0000 7089 a2f1 008f
0001090 0000 00f1 0091 0000 00f1 0090 0000 00f6
00010a0 008e 0000 00f8 008f 0000 00f7 008f 0000
00010b0 0000 0000 0000 0000 0000 0000 0000 0000
00010c0 0000 0070 ac70 89ac 7089 ac70 89ac 3636
00010d0 3430 3636 3430 3430 2824 3430 2824 3430
00010e0 2824 3434 2e28 2420 2e28 2420 2e28 2420
00010f0 3434 2e28 2420 2e28 2420 2e28 2420 3434
0001100 2e28 2420 2e28 2420 2e28 2420 3430 2c28
0001110 2420 2c28 2420 2c28 2420 3430 2c28 2420
0001120 2c28 2420 2c28 2420 3430 2c28 2420 2c28
0001130 2420 2c28 2420 1112 1517 4142 4547 3132
0001140 3537 7075 9da2 7075 a2ff 7075 a2ff 7a7f
0001150 9398 7075 acb8 7075 ac00 7075 ac00 7a7f
0001160 93a2 7075 ac00 7075 ac00 7075 ac00 7a7f
0001170 93a2 3c7c 3c3c 3c7c 3c3c 7c3c 3c7c 7c3c
0001180 3c3c 3c7c 3c3c 3c7c 3c3c 3c7c 7c3c 3c7c
0001190 3c3c 3c7c 3c3c 3c7c 3c3c 3c7c 7c7c 3c7c
00011a0 7c7c 1001 0000 2222 0200 0000 0000 0000
00011b0 0000 0000 0000 4400 0000 0000 00ff 0000
00011c0 0000 0000 0000 0000 0000 0000 000e 0e03
00011d0 002d e200 020e 1ce0 e0f0 0ce0 e0f0 6c00
00011e0 0000 0000 0000 0000 0013 0000 0000 0000
00011f0 0000 0000 0000 004c 5468 788c a0b9 cd00
0001200 0000 0000 0000 0000 0000 0000 0000 0000
*
0001280 0000 0000 0000 0000 0000 0000 0000 004c
0001290 5468 788c a0b9 cd4c 5868 8cb4 bdb9 cd4c
00012a0 5868 8cb4 bdb9 cd14 1414 0a14 1414 0a14
00012b0 1414 0a14 1414 0a14 1414 0a14 1414 0a14
00012c0 1414 0a14 1414 0a14 140a 0a00 000a 0a00
00012d0 000a 0a00 0014 140a 0a00 000a 0a00 000a
00012e0 0a00 0014 140a 0a00 000a 0a00 000a 0a00
00012f0 0014 140a 0a00 000a 0a00 000a 0a00 0014
0001300 140a 0a00 000a 0a00 000a 0a00 0014 140a
0001310 0a00 000a 0a00 000a 0a00 0014 140a 0a00
0001320 000a 0a00 000a 0a00 0014 140a 0a00 000a
0001330 0a00 000a 0a00 0014 140a 0a00 000a 0a00
0001340 000a 0a00 0014 140a 0a00 000a 0a00 000a
0001350 0a00 0014 140a 0a00 000a 0a00 000a 0a00
0001360 0014 140a 0a00 000a 0a00 000a 0a00 0014
0001370 140a 0a00 000a 0a00 000a 0a00 0014 140a
0001380 0a00 000a 0a00 000a 0a00 0014 140a 0a00
0001390 000a 0a00 000a 0a00 0014 140a 0a00 000a
00013a0 0a00 000a 0a00 0010 1618 4046 4830 3638
00013b0 4c5c 608c a0b4 bdcd 4c5c 608c 90b4 bdcd
00013c0 4e56 5e66 8e96 aebf 4c50 5c68 8cb4 ffff
00013d0 4c5c 8cb4 ffff ffff 4e5e 668e 9eae ffff
00013e0 4c50 545c 8ca0 b4bd 4c5c 688c 98b4 bdcd
00013f0 4e56 5e8e 96ae bfc7 7c7c 7c7c 7c7c 7c3c
0001400 7c7c 7c7c 7c7c 7c3c 3c7c 3c7c 7c7c 7c7c
0001410 3c7c 7c3c 7c3c 3c3c 7c7c 7c3c 3c3c 3c3c
0001420 7c7c 7c7c 7c3c 3c3c 7c7c 7c7c 7c7c 7c7c
0001430 7c7c 3c7c 7c7c 7c3c 7c3c 7c7c 7c7c 3c7c
0001440 ffff ffff ffff ffff ffff ffff ffff ffff
*
0010000